crates/cubecl-wgpu/src/compiler/base.rs

```diff
@@ -1,31 +1,113 @@
-use std::sync::Arc;
+use std::fmt::Display;
 
 use cubecl_common::ExecutionMode;
-use cubecl_core::{prelude::CompiledKernel, server::ComputeServer, Compiler, Feature};
-use cubecl_runtime::DeviceProperties;
-use wgpu::{Adapter, ComputePipeline, Device, Queue};
+use cubecl_core::{
+    prelude::{CompiledKernel, KernelDefinition},
+    server::ComputeServer,
+    Compiler, CompilerRepresentation, WgpuCompilationOptions,
+};
+use derive_more::derive::From;
 
-use crate::WgpuServer;
+use crate::{WgpuServer, WgslCompiler};
+
+use super::wgsl::ComputeShader;
+
+#[derive(Debug, Clone)]
+pub enum DynCompiler {
+    Wgsl(WgslCompiler),
+    #[cfg(feature = "spirv")]
+    SpirV(cubecl_spirv::SpirvCompiler),
+}
+
+#[derive(From)]
+#[allow(clippy::large_enum_variant)]
+pub enum DynRepresentation {
+    Wgsl(ComputeShader),
+    #[cfg(feature = "spirv")]
+    SpirV(cubecl_spirv::SpirvKernel),
+}
+
+#[cfg(feature = "spirv")]
+impl DynRepresentation {
+    pub fn as_spirv(&self) -> Option<&cubecl_spirv::SpirvKernel> {
+        match self {
+            DynRepresentation::SpirV(repr) => Some(repr),
+            _ => None,
+        }
+    }
+}
+
+impl Display for DynRepresentation {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            DynRepresentation::Wgsl(compute_shader) => compute_shader.fmt(f),
+            #[cfg(feature = "spirv")]
+            DynRepresentation::SpirV(spirv_kernel) => spirv_kernel.fmt(f),
+        }
+    }
+}
+
+impl CompilerRepresentation for DynRepresentation {
+    fn shared_memory_size(&self) -> usize {
+        0
+    }
+}
```

Comment:
Is this a problem? I guess we should do a match and call `shared_memory_size` on the inner compiler?

Reply:
It was never used in anything other than CUDA, I think it should actually just be on the CUDA repr itself now that we give access to that